### PR TITLE
 SONARPY-1199 Avoid scanning files that don't need to be analyzed 

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**

### CODE REVIEW

**Review by gyula-sallai-sonarsource — CHANGES_REQUESTED** [review]
Very nice job! I've left a couple of comments, but there is no blocker. I am putting the ticket back into progress, because given the size of the PR, it would be nice to have another quick round of review.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexer.java [hunk 1]
```diff
@@ -19,30 +19,129 @@
...
   }
 
   @Override
   public void buildOnce(SensorContext context) {
     this.projectBaseDirAbsolutePath = context.fileSystem().baseDir().getAbsolutePath();
-    PerformanceMeasure.Duration duration = PerformanceMeasure.start("ProjectLevelSymbolTable");
-    LOG.debug("Input files for indexing: " + files);
-    // computes "globalSymbolsByModuleName"
+    LOG.debug("Input files for indexing: " + mainFiles);
+    boolean shouldOptimizeAnalysis = context.config().getBoolean(SONAR_CAN_SKIP_UNCHANGED_FILES_KEY).orElse(false) && cacheContext.isCacheEnabled();
```
[the remark this answers is not in the document]
Comment (gyula-sallai-sonarsource) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexer.java]: This comment is outdated, feel free to discard it.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexer.java [hunk 2]
```diff
@@ -19,30 +19,133 @@
...
+    computeGlobalSymbolsUsingCache(context);
+  }
+
+  private void computeGlobalSymbolsUsingCache(SensorContext context) {
+    LOG.info("Using cached data to retrieve global symbols.");
+    // TODO SONARPY-1196: Compute deleted files here through diff with what was saved and include them to projectModuleFQNs / modifiedFiles
+    Set<String> projectModulesFQNs = new HashSet<>(inputFileToFQN.values());
+    Map<String, Set<String>> importsByModule = new HashMap<>();
+    List<InputFile> modifiedFiles = new ArrayList<>();
+    List<String> impactfulFilesFQN = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexer.java]: As discussed I'll rename the first list to `impactfulFiles` which is more consistent which how it's used and that should already help with the confusion, especially thanks to your other refactoring suggestions.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexer.java [hunk 3]
```diff
@@ -19,30 +19,133 @@
...
+      }
+      Set<Descriptor> descriptors = caching.readProjectLevelSymbolTableEntry(inputFile.key());
+      if (descriptors != null && imports != null) {
+        saveRetrievedDescriptors(inputFile.key(), descriptors, caching);
+      } else {
+        // Failed to retrieve some data: consider the file as impactful.
+        impactfulFilesFQN.add(currFQN);
+        modifiedFiles.add(inputFile);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexer.java]: Very nice suggestion, thanks! That makes the code much cleaner.

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexerTest.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+  public void test_modified_dependency() {
+    file1 = createInputFile(baseDir, "main.py", InputFile.Status.SAME, InputFile.Type.MAIN);
+    file2 = createInputFile(baseDir, "mod.py", InputFile.Status.CHANGED, InputFile.Type.MAIN);
+
+    List<InputFile> inputFiles = new ArrayList<>(Arrays.asList(file1, file2));
+    moduleFileSystem = new TestModuleFileSystem(inputFiles);
+
+    byte[] serializedSymbolTable = toProtobufModuleDescriptor(Set.of(new VariableDescriptor("x", "main.x", null))).toByteArray();
+    byte[] outdatedEntry = toProtobufModuleDescriptor(Set.of(new VariableDescriptor("outdated", "mod.outdated", null))).toByteArray();
+    readCache.put(IMPORTS_MAP_CACHE_KEY_PREFIX + "moduleKey:main.py", String.join(";", List.of("mod")).getBytes(StandardCharsets.UTF_8));
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/test/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexerTest.java]: I didn't add an additional test, but I added some unresolved imports surrounding the `mod` one being tested. Please let me know if you would like to see something more thorough!

**Review by gyula-sallai-sonarsource — APPROVED** [review]
Looks good to me!

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java [hunk 5]
```diff
@@ -105,14 +109,19 @@ public void describe(SensorDescriptor descriptor) {
...
       analysisWarnings.addUnique(UNSET_VERSION_WARNING);
     }
     pythonVersionParameter.ifPresent(value -> ProjectPythonVersion.setCurrentVersions(PythonVersionUtils.fromString(value)));
-    PythonIndexer pythonIndexer = this.indexer != null ? this.indexer : new SonarQubePythonIndexer(mainFiles);
+    CacheContext cacheContext;
+    if (!context.runtime().getProduct().equals(SonarProduct.SONARLINT) && context.runtime().getApiVersion().isGreaterThanOrEqual(Version.create(9, 7))) {
+      cacheContext = new CacheContextImpl(context.isCacheEnabled(), new PythonWriteCacheImpl(context.nextCache()), new PythonReadCacheImpl(context.previousCache()));
+    } else {
+      cacheContext = new CacheContextImpl(false, new DummyCache(), new DummyCache());
+    }
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java]: Did that on implementation side with a `CacheContextImpl#of` instead, but indeed that cleans things up a bit!

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexerTest.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+  public LogTester logTester = new LogTester();
+
+  InputFile file1;
+  InputFile file2;
+  TestModuleFileSystem moduleFileSystem;
+  SonarQubePythonIndexer pythonIndexer;
+  ProjectLevelSymbolTable projectLevelSymbolTable;
+  TestReadCache readCache;
+  TestWriteCache writeCache;
+  CacheContextImpl cacheContext;
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/test/java/org/sonar/plugins/python/indexer/SonarQubePythonIndexerTest.java]: Yes, and actually doing that made me realize I didn't need some of these (the joy of copy pasting stuff 😅).

**Pull request merged**